llvm/test/Transforms/SCCP/loop-removal.ll

```diff
@@ -0,0 +1,43 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 5
+; RUN: opt < %s -passes=ipsccp -S | FileCheck %s
+
+define i32 @foo() {
+; CHECK-LABEL: @foo(
+; CHECK-NEXT:  init:
+; CHECK-NEXT:    br label %[[OUTER_LOOP_CONTROL:.*]]
+; CHECK:       outer.loop.control:
+; CHECK-NEXT:    [[X_0:%.*]] = phi i32 [ 0, [[INIT:%.*]] ], [ [[X_OUTER:%.*]], [[OUTER_LOOP_INC:%.*]] ]
+; CHECK-NEXT:    [[TMP0:%.*]] = icmp slt i32 [[X_0]], 10
+; CHECK-NEXT:    br i1 [[TMP0]], label %[[INNER_LOOP_CONTROL:.*]], label %[[EXIT:.*]]
+; CHECK:       inner.loop.control:
+; CHECK-NEXT:    br label [[OUTER_LOOP_INC]]
+; CHECK:       outer.loop.inc:
+; CHECK-NEXT:    [[X_OUTER]] = add nsw i32 [[X_0]], 2
+; CHECK-NEXT:    br label %[[OUTER_LOOP_CONTROL]]
+; CHECK:       exit:
+; CHECK-NEXT:    ret i32 [[X_0]]
+;
+init:
+  br label %outer.loop.control
+
+outer.loop.control:                                                ; preds = %init, %outer.loop.inc
+  %x.0 = phi i32 [ 0, %init ], [ %x.outer, %outer.loop.inc ]
+  %0 = icmp slt i32 %x.0, 10
+  br i1 %0, label %inner.loop.control, label %exit
+
+inner.loop.control:                                                ; preds = %outer.loop.control, %inner.loop.body
+  %x.1 = phi i32 [ %x.0, %outer.loop.control ], [ %x.inner, %inner.loop.body ]
+  %1 = icmp sgt i32 %x.1, 20
```

Comment:
I think we should eliminate this comparison in ConstraintElimination.

The constraint system
```
%x.0 s< 10
%x.1 s<= %x.0 (%x.1 is monotonically decreasing)
%x.1 s> 20
```
is unsatisfiable.

Reply:
Thank you for your suggestion. I was under the impression that ConstraintElimination pass cannot handle loops well. I will take a closer look at ConstraintElimination pass.

Reply:
Personally I don't want to pay for the complexity to handle random-generated cases :(
cc @nikic for more guidance.

Reply:
@dtcxzyw -- this can happen (and we saw it happening) in real code. Imagine a round of inlining followed by function passes that leaves some dead code behind.

Reply:
> @dtcxzyw -- this can happen (and we saw it happening) in real code. Imagine a round of inlining followed by function passes that leaves some dead code behind.

Also GCC seems to get this one right, so the optimization is not that far-fetched.
https://godbolt.org/z/rMMdTsr7f

Reply:
> this can happen (and we saw it happening) in real code. Imagine a round of inlining followed by function passes that leaves some dead code behind.

Please provide code pieces from some real-world projects.

Reply:
It's reduced from a piece of source that I can't, unfortunately, share.

Reply:
> > @dtcxzyw -- this can happen (and we saw it happening) in real code. Imagine a round of inlining followed by function passes that leaves some dead code behind.
> 
> Also GCC seems to get this one right, so the optimization is not that far-fetched. https://godbolt.org/z/rMMdTsr7f

See https://github.com/llvm/llvm-project/issues/70985#issuecomment-1790563680. I don't buy the argument that "GCC does this optimization but llvm doesn't." Actually I don't have strong objection to optimize random-generated cases **if we can do this with small code changes**.

Reply:
I understand and appreciate your comment -- I think, if anything, the testcase shows a hole/gap in the range analysis infra in LLVM. I don't have a strong opinion on how we should fix it.

Reply:
> But in the original issue (https://github.com/llvm/llvm-project/issues/108906) nested while loops are not SCEVable.

@grigorypas IIRC ConstraintElimination only handles monotonically increasing indvars. It would be easy to support monotonically decreasing cases.

Reply:
Regarding https://github.com/llvm/llvm-project/issues/108906, I agree that we should not try to optimize such cases. If I ever find the time, I'll write a new section for the developer policy that specifies requirements for fuzzer-generated issue.

As for the change itself, I think the idea behind it is fairly reasonable, and I expect that it will help more realistic cases as well. Unfortunately, we're currently pretty bad with bounds checks elimination in loops for a combination of different reasons (SCEV being glacially slow with context sensitive facts, LVI going to overdefined in cycles, ConstraintElimination only having very basic loop reasoning, etc).

I see some code size changes from this patch: http://llvm-compile-time-tracker.com/compare.php?from=eaea5f6f952b6059cebfe87ea9800a3a6516f9ed&to=6358352b3b4a9912e4dd312fad057b02833a7faa&stat=size-text And also significant compile-time regressions, so we'll have to see if we can make it cheap enough: http://llvm-compile-time-tracker.com/compare.php?from=eaea5f6f952b6059cebfe87ea9800a3a6516f9ed&to=6358352b3b4a9912e4dd312fad057b02833a7faa&stat=instructions%3Au I haven't had time to look at the implementation yet.

Reply:
> Regarding https://github.com/llvm/llvm-project/issues/108906, I agree that we should not try to optimize such cases. If I ever find the time, I'll write a new section for the developer policy that specifies requirements for fuzzer-generated issue.

While I agree that LLVM shouldn't try to optimize crazy cases, I still do believe the code in #108906 is not that unreasonable to optimize. I found out several instances in real codebases where a line of dead code was preventing other optimizations to kick in, causing code bloat.

Reply:
Also, on a more general note, I think a range analysis infra shouldn't necessarily be concerned with how the code looks like as input, as long as it has reasonable capabilities to reason about equalities and inequalities. I looked at some point and found out the limitation you point out (for LVI/CE), and I wonder if instead of fixing these problems one by one we might want to rethink the bigger picture.